### ISSUE TITLE
Race condition: DsWriteThenArchiver may start while put call is in process

### ISSUE THREAD

**Issue by RKrahl** [icatproject/ids.server#47]
There is a race condition in ids.server: it may happen that the DsWriteThenArchiver is started for a dataset while an upload of a new file to the same dataset via the put webservice call is in process.  In the worst case this may lead to a complete loss of all data in this dataset.

This scenario assumes the ids server to be configured as two level storage with storage unit set to dataset.  I reproduced the issue with the current release version ids.server 1.5.0 / ids.storage_file 1.3.2 as well as with the snapshot ids.server 1.6.0-20160309.101441-6 / ids.storage_file 1.3.3-20160307.153059-1.

Steps to reproduce:

1. Upload a file to the dataset.  The file should be fairly large to make sure that writing the archive file in step 4 takes some time.
2. Concurrently, start a slow upload of a file to the same dataset.
3. Wait for the first upload to be finished.  (The FSM will queue a WRITE deferred operation.)  Call archive for the dataset.  This will change the WRITE operation to WRITE_THEN_ARCHIVE.
4. Wait a little longer then 60 seconds (assuming this is the value of writeDelaySeconds in ids.properties).  Now the DsWriteThenArchiver thread will be started.
5. Finish the slow upload from step 2.  This queues another WRITE deferred operation.
6. The DsWriteThenArchiver thread will delete the dataset from main storage, including the newly uploaded datafile, that has not been included in the archive file.  Another 60 seconds after step 5, the FSM starts a DsWriter thread.  Since the dataset is not in main storage any more, it will delete the ZIP file from archive storage.
   -> Boom!  All data is lost.

I prepared a Python script [race_upload_archive.py](https://github.com/icatproject/ids.server/files/197105/race_upload_archive.py.txt) that reproduces the issue.  There is also a [ids.log](https://github.com/icatproject/ids.server/files/197107/race_upload_archive.log.txt) file.

**Comment by RKrahl** [comment]
After having spent some time reading and analyzing the sources, I'd like to add a more general comment:

While starting threads to perform deferred operations, the FiniteStateMachine takes care to guarantee that never more then one thread may operate on the same dataset at any time.  But some webservice calls also operate directly on the storage.  As the calls come in unpredictably, conflicts may arise between these calls and background threads.  There is some protection, but this is not perfect, as this issue shows.

I found other possible cases of concurrent and sometimes conflicting access to the same dataset in main storage, but most of them seem to be mostly harmless.  The present issue was the only case I found that actually causes significant damage.  But things are complicated and I'm not sure to have checked all possibilities.

There are different mechanisms in place to prevent conflicting storage access.  My impression from reading the sources was that these mechanisms are already too complicated but still not complete.  A review and simplification of the locking in ids.server might be advisable.

**Comment by RKrahl** [comment]
Concerning the present issue, the cause is the following: the put webservice call sets an ARCHIVE lock on the dataset.  This prevents the FiniteStateMachine to start a DsArchiver thread before the upload is complete.  So far so good.  But the FSM does not check the ARCHIVE lock before starting a DsWriteThenArchiver and that is the problem here.

On the other hand, the FSM does not postpone a queued ARCHIVE_REQUESTED operation if an ARCHIVE lock is set, but rather drops it completely (is this intended btw?).  Of course it could not simply drop a WRITE_THEN_ARCHIVE_REQUESTED operation as at least the write part must not be ignored.

**Comment by fisherab** [comment]
Thanks for the analysis. I think the simplest solution is to change the FSM code inside the DsProcessQueue in the block starting at line 165. If the ARCHIVE lock is set then simply instantiate a DsWriter rather than DsWriteThenArchiver. In general I don't mind silently ignoring archive requests as I expect most archive requests to be generated internally when there is insufficient space available.

**Comment by RKrahl** [comment]
Why adding yet another special condition?  Why don't you use my PR which removes a large chunk of redundant code and thus significantly simplifies the code?

**Comment by fisherab** [comment]
I will study your pull request. If it removes code I will be happy with it!

**Comment by fisherab** [comment]
I have now looked at the code and so far I don't see that it introduces new problems. It could be argued that it is conceptually slightly more complex because of the extra newOps map however it does save a lot of code. I thought I needed writethenarchive to make sure that files would not be removed from the cache if the writing fails. This half works - however with the existing code (or my previous proposed modification) if a write or writethenarchive fails in the write step any subsequent archive will be honoured resulting in deletion of data in the cache which has not been pushed to archive storage. I already have the marker files so that if the IDS crashes or is forcibly restarted then the missed writes will be executed. I propose to accept your pull request and also to ensure that the archiver skips it's action if the marker file still exists.
